cmd/nsdns/watch.go

```diff
@@ -0,0 +1,192 @@
+			}
+
+			api := namesilo_api.NewNamesiloApi(domainName, nsApiKey)
+
+			rrMutex := sync.RWMutex{}
+			var records []namesilo_api.ResourceRecord
+			var ip string
+
+			refreshState := func() error {
+				var err error
+				rrMutex.Lock()
+				defer rrMutex.Unlock()
+
+				records, err = api.ListDNSRecords()
+				if err != nil {
+					return err
+				}
+				ip, err = icanhazip.GetPublicIP()
+				if err != nil {
+					return err
+				}
+
+				log.Debug("Updated local resource records and IP address")
+
+				return nil
+			}
+
+			if err := refreshState(); err != nil {
+				return err
+			}
+
+			go func() {
+				for range time.Tick(time.Hour) {
+					if err := refreshState(); err != nil {
+						panic(err)
+					}
+				}
+			}()
+
+			clientset, err := GetKubernetesClientSet()
+			if err != nil {
+				return err
+			}
+
+			informerFactory := informers.NewSharedInformerFactory(clientset, time.Minute)
+
+			informerFactory.Networking().V1().Ingresses().Informer().AddEventHandler(
+				cache.ResourceEventHandlerFuncs{
+					AddFunc: func(obj interface{}) {
+						ingress := obj.(*networkingv1.Ingress)
+						if !ShouldProcessIngress(ingressClass, ingress) {
+							return
+						}
+
+						record, err := nsdns.NamesiloRecordFromIngress(ingress, domainName, ip)
+						if err != nil {
+							log.Error(err)
+							return
+						}
+
+						existingRecord, err := RecordMatching(records, *record)
+						if existingRecord != nil {
+							return
+						}
+
+						log.Infof("Adding record for %s", record.Host)
+						if err := api.AddDNSRecord(*record); err != nil {
+							log.Error(err)
+						}
+
+						if err := refreshState(); err != nil {
+							log.Error(err)
+						}
+					},
+					DeleteFunc: func(obj interface{}) {
+						ingress := obj.(*networkingv1.Ingress)
+						if !ShouldProcessIngress(ingressClass, ingress) {
+							return
+						}
+
+						record, err := nsdns.NamesiloRecordFromIngress(ingress, domainName, ip)
+						if err != nil {
+							log.Error(err)
+							return
+						}
+
+						deleteRecord, err := RecordMatching(records, *record)
+
+						log.Infof("Deleting resource record %s", deleteRecord.RecordId)
+						err = api.DeleteDNSRecord(*deleteRecord)
+						if err != nil {
+							log.Error(err)
+							return
+						}
+
+						if err := refreshState(); err != nil {
+							log.Error(err)
+						}
+					},
+					UpdateFunc: func(old, new interface{}) {
+						ingress := new.(*networkingv1.Ingress)
+						if !ShouldProcessIngress(ingressClass, ingress) {
+							return
+						}
+
+						record, err := nsdns.NamesiloRecordFromIngress(ingress, domainName, ip)
+						if err != nil {
+							log.Error(err)
+							return
+						}
+
+						// Only update if something actionable changed.
+						updateRecord, err := RecordMatching(records, *record)
+						if record.Equals(updateRecord) {
+							return
+						}
+
+						if err := api.UpdateDNSRecord(*updateRecord); err != nil {
```

Comment:
Should add a log above this line, because right now it looks like updates are complete silent.